### PR TITLE
Add fp8 custom op and unit test

### PR DESCRIPTION
**Pull request by wenscarl**
In this PR, we add one custom op to utilize the fp8 gemms provided by the XLA to better utilize NVIDIA Hopper GPUs. Users can then code-inject this custom op like:
```
dense=nn.DenseGeneral(dot_general_cls=nn.Fp8EinsumOp)
```
The PR also provides unit tests for the injection and accuracy.

Reference:

[FP8 basics](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html).

[Basics of QDQ process used in our custom op](https://github.com/NVIDIA/atex/blob/experimental-xla-fp8-layers/experimental/xla-fp8-layers/fp8-tutorial.md).

cc. @pjannaty @reedwm @kaixih @levskaya

### CODE REVIEW

File: tests/linen/linen_linear_test.py [hunk 1]
```diff
@@ -1016,6 +1017,66 @@ def __call__(self, x):
...
+
+      def _train(variables, x):
+        y = p.apply(variables, x)
+        loss = y * dy
+        return jnp.mean(loss)
+      train_fn = jax.jit(jax.value_and_grad(_train, argnums=[0, 1]))
+      outputs, grads = train_fn(initial_vars, x)
+      return outputs, grads
+     
+    #y, variables = Foo().init_with_output(random.PRNGKey(0), x)
```
Comment (kaixih) [inline tests/linen/linen_linear_test.py:1057]: Can we remove the commented lines?

File: flax/linen/fp8_ops.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+variable_with_axes = nn_partitioning.variable_with_axes
+
+# Type annotations
+Array = jnp.ndarray
+Dtype = jnp.dtype
+PRNGKey = jnp.ndarray
+Shape = Iterable[int]
+ActivationFn = Callable[..., Array]
+DotGeneralT = Callable[..., Array]
+PrecisionLike = Union[
```
Comment (kaixih) [inline flax/linen/fp8_ops.py:26]: Please clean up the unused alias and imports.

File: tests/linen/linen_linear_test.py [hunk 3]
```diff
@@ -1016,6 +1017,66 @@ def __call__(self, x):
     )
     self.assertEqual(y.shape, (2, 8, 6))
 
+  def test_fp8_dot_general_cls_injection(self):
+    # Used to cast the inputs to be representable in FP8, so that the difference
+    # of the results from the original gemm and fp8 gemm is small.
+    cast_to_representable = functools.partial(nn.fp8_quantize_dequantize,
+                                              scale=jnp.ones((1,)),
+                                              compute_dtype=jnp.float32)
+    class Foo(nn.Module):
```
Comment (kaixih) [inline tests/linen/linen_linear_test.py:1026]: Why do we need a Foo class? Can we directly use the DenseGeneral and provide our custom op to it?

**Review by kaixih — COMMENTED** [review]
Thanks. Just added some minor comments.

File: flax/linen/fp8_ops.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+from flax.core import scope as flax_scope
+from flax.linen import initializers
+from flax.linen.module import Module
+from flax.linen import partitioning as nn_partitioning
+from jax import custom_vjp
+from jax import lax
+from jax import numpy as jnp
+from jax import random
+
+from fp8layers.jax_common.fp8 import fp8_dot_general
```
Comment (kaixih) [inline flax/linen/fp8_ops.py:14]: This will cause module importing error and needs to be removed. Please clean up other unused imports.

File: flax/linen/fp8_ops.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        jnp.float32,
+    )
+    amax_history_args = (
+        initializers.zeros_init(),
+        random.PRNGKey(0),
+        (self.amax_history_length,),
+        jnp.float32,
+    )
+
+    self.input_amax_history = variable_with_axes(
```
Comment (levskaya) [inline flax/linen/fp8_ops.py:143]: ...why are you using the deprecated paritioning helper `variable_with_axes` to seed a trivial piece of metadata here?  Why not just use `self.variable`?  I don't see anything in this code that's actually doing any non-trivial operation using this metadata.

File: flax/training/train_state.py [hunk 6]
```diff
@@ -15,10 +15,23 @@
...
+  axis_names = nn.partitioning.get_axis_names(params_axes)
+  missing_params_axes = (
+      set(traverse_util.flatten_dict(params, sep='/')) -
+      set(traverse_util.flatten_dict(axis_names, sep='/')))
+  if missing_params_axes:
+    raise ValueError(
+        f'Missing axis names for parameters: {missing_params_axes}')
+
+
 class TrainState(struct.PyTreeNode):
```
Comment (levskaya) [inline flax/training/train_state.py:35]: I might ask that we add this FP8 specific logic to a subclass of the general TrainState class? (it can be located in this file).

**Review by levskaya — COMMENTED** [review]
I can help advise how to replace it, but definitely don't use the old `variable_with_axes`.  I'm not sure you even need it for what you're trying to do here.  e.g. If you just want to mark the fp8 use-places we can use a simple flax.struct.dataclass with a single-field as a kind of leaf type-marker.

File: flax/linen/fp8_ops.py (hunk 5 above)
Comment (kaixih) [reply flax/linen/fp8_ops.py:143]: Sure. I think we can simply use the `self.variable`. The reason why we used the `variable_with_axes`  was that we originally developed this in the context of t5x, where it requires all the variables have the axes information. Since this PR focuses on the FLAX, let's remove them. @wenscarl

Comment (wenscarl) [reply flax/linen/fp8_ops.py]: `variable_with_axes` has been replaced by `self.params`.

File: tests/linen/linen_linear_test.py [hunk 7]
```diff
@@ -1016,6 +1020,142 @@ def __call__(self, x):
...
+
+  def test_fp8_with_train_state(self):
+    x = random.uniform(random.PRNGKey(1), (16, 16), dtype=jnp.float32)
+    dense = nn.DenseGeneral(features=32, use_bias=True,
+                            dot_general_cls=nn.Fp8DenseGeneralOp)
+    key = random.PRNGKey(0)
+    variables = dense.init(key, x)
+
+    opt = optax.adam(learning_rate=.1)
+    state = train_state.TrainState.create(params=variables, tx=opt,
```
Comment (levskaya) [inline tests/linen/linen_linear_test.py:1084]: shouldn't this be `Fp8TrainState` now?

**Review by levskaya — APPROVED** [review]
Looks good - just one question inline about your test code:

**Comment by levskaya** [comment]
bleh sorry it looks like pyink formatter is complaining again (we're about to disable it for a while as this is proving more annoying than helpful - sorry about that.)

**Comment by levskaya** [comment]
We disabled the autoformatter stuff, I tried retriggering tests, but it's not capturing the update, so if you rebase and push a new commit hopefully it will clear.  Sorry for the trouble.

File: tests/linen/linen_linear_test.py (hunk 7 above)
Comment (wenscarl) [reply tests/linen/linen_linear_test.py]: Good catch and fixed. Thanks!
